File: Sample/sample_cpp/ObjectSystem/GeometryObject2D.cpp
#include <Base.h>

/**
@brief	図形を表示する。
*/
void GeometryObject2D()
{
	// aceを初期化する
	ace::Engine::Initialize(ace::ToAString("GeometryObject2D").c_str(), 640, 480, ace::EngineOption());

	// シーンを生成する
	auto scene = std::make_shared<ace::Scene>();

	auto layer = std::make_shared<ace::Layer2D>();
	auto texture = ace::Engine::GetGraphics()->CreateTexture2D(ace::ToAString("Data/Texture/Sample1.png").c_str());

	//図形描画クラスのコンストラクタを呼び出す
	auto geometryObj1 = std::make_shared<ace::GeometryObject2D>();
	auto geometryObj2 = std::make_shared<ace::GeometryObject2D>();
	auto geometryObj3 = std::make_shared<ace::GeometryObject2D>();
	auto geometryObj4 = std::make_shared<ace::GeometryObject2D>();
	auto geometryObj5 = std::make_shared<ace::GeometryObject2D>();
	auto geometryObj6 = std::make_shared<ace::GeometryObject2D>();

	//図形描画クラスをレイヤーに追加する。
	layer->AddObject(geometryObj1);
	layer->AddObject(geometryObj2);
	layer->AddObject(geometryObj3);
	layer->AddObject(geometryObj4);
	layer->AddObject(geometryObj5);
	layer->AddObject(geometryObj6);

	//レイヤーをシーンに追加する。
	scene->AddLayer(layer);

	//シーンを変更する。
	ace::Engine::ChangeScene(scene);

	//円を図形描画クラス1にて描画する。
	{
		//円の外径、内径、頂点数、中心位置を指定。
		auto circle = std::make_shared<ace::CircleShape>();
		circle->SetOuterDiameter(100);
		circle->SetInnerDiameter(10);
		circle->SetNumberOfCorners(96);
		circle->SetPosition(ace::Vector2DF(100, 50));

		//円を描画する図形として設定し、合成するテクスチャも設定。
		geometryObj1->SetShape(circle);
		geometryObj1->SetTexture(texture);
		geometryObj1->SetPosition(ace::Vector2DF(0, 0));
	}

	//弧を図形描画クラス2にて描画する。
		{
			//弧の外径、内径、頂点数、中心位置、開始頂点番号、終了頂点番号を指定。
			auto arc = std::make_shared<ace::ArcShape>();
			arc->SetOuterDiameter(100);
			arc->SetInnerDiameter(10);
			arc->SetNumberOfCorners(96);
			arc->SetPosition(ace::Vector2DF(300, 50));
			arc->SetStartingCorner(90);
			arc->SetEndingCorner(5);

			//弧を描画する図形として設定し、合成するテクスチャも設定。
			geometryObj2->SetShape(arc);
			geometryObj2->SetTexture(texture);
			geometryObj2->SetPosition(ace::Vector2DF(0, 0));
		}

		//線分を図形描画クラス3にて描画する。
		{
			//線分の始点、終点、太さを指定。
			auto line = std::make_shared<ace::LineShape>();
			line->SetStartingPosition(ace::Vector2DF(410, 50));
			line->SetEndingPosition(ace::Vector2DF(630, 50));
			line->SetThickness(5.0f);

			//線分を描画する図形として設定。
			geometryObj3->SetShape(line);
			geometryObj3->SetPosition(ace::Vector2DF(0, 0));
		}

		//矩形を図形描画クラス4にて描画する。
		{
			//矩形の描画範囲、UV範囲を指定。
			auto rect = std::make_shared<ace::RectangleShape>();
			rect->SetDrawingArea(ace::RectF(10, 110, 300, 200));
			rect->SetUV(ace::RectF(0, 0, 0.5, 0.5));

			//矩形を描画する図形として設定し、合成するテクスチャも設定。
			geometryObj4->SetShape(rect);
			geometryObj4->SetTexture(texture);
			geometryObj4->SetPosition(ace::Vector2DF(0, 0));
		}

		//三角形を図形描画クラス5にて描画する。
		{
			//三角形の各頂点の位置とUVを指定。
			auto triangle = std::make_shared<ace::TriangleShape>();
			triangle->SetPointByIndex(ace::Vector2DF(320, 350), 0);
			triangle->SetPointByIndex(ace::Vector2DF(100, 450), 1);
			triangle->SetPointByIndex(ace::Vector2DF(540, 450), 2);

			triangle->SetUVByIndex(ace::Vector2DF(0.5, 0.2), 0);
			triangle->SetUVByIndex(ace::Vector2DF(0.1, 0.5), 1);
			triangle->SetUVByIndex(ace::Vector2DF(0.9, 0.7), 2);

			//三角形を描画する図形として設定し、合成するテクスチャも設定。
			geometryObj5->SetShape(triangle);
			geometryObj5->SetTexture(texture);
			geometryObj5->SetPosition(ace::Vector2DF(0, 0));
		}

		//多角形を図形描画クラス6にて描画する。
		{

			auto polygon = std::make_shared<ace::PolygonShape>();
			//多角形を構成する頂点を追加していく。（星形になるようにする。）
			for (int i = 0; i < 10; ++i)
			{
				ace::Vector2DF vec = ace::Vector2DF(1, 0);
				vec.SetDegree(i * 36);
				vec.SetLength((i % 2 == 0) ? 100 : 55);
				polygon->AddVertex(vec + ace::Vector2DF(500, 250));

			}

			//多角形を描画する図形として設定し、合成するテクスチャも設定。
			geometryObj6->SetShape(polygon);
			geometryObj6->SetTexture(texture);
			geometryObj6->SetPosition(ace::Vector2DF(0, 0));
		}

		// aceが進行可能かチェックする。
		while (ace::Engine::DoEvents())
		{
			// aceを更新する。
			ace::Engine::Update();
		}

		// aceを終了する。
		ace::Engine::Terminate();
}

File: Sample/sample_cs/Graphics/2D/GeometryObject2D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace sample_cs.Graphics._2D
{
    /// <summary>
    /// 図形を表示する。
    /// </summary>
    class GeometryObject2D : ISample
    {
        public void Run()
        {
            // aceを初期化する
            ace.Engine.Initialize("GeometryObject2D", 640, 480, new ace.EngineOption());

            // シーンを生成する
            var scene = new ace.Scene();

            var layer = new ace.Layer2D();
            var texture = ace.Engine.Graphics.CreateTexture2D("Data/Texture/Sample1.png");

            //図形描画クラスのコンストラクタを呼び出す
            var geometryObj1 = new ace.GeometryObject2D();
            var geometryObj2 = new ace.GeometryObject2D();
            var geometryObj3 = new ace.GeometryObject2D();
            var geometryObj4 = new ace.GeometryObject2D();
            var geometryObj5 = new ace.GeometryObject2D();
            var geometryObj6 = new ace.GeometryObject2D();

            //図形描画クラスをレイヤーに追加する。
            layer.AddObject(geometryObj1);
            layer.AddObject(geometryObj2);
            layer.AddObject(geometryObj3);
            layer.AddObject(geometryObj4);
            layer.AddObject(geometryObj5);
            layer.AddObject(geometryObj6);

            //レイヤーをシーンに追加する。
            scene.AddLayer(layer);

            //シーンを変更する。
            ace.Engine.ChangeScene(scene);

            //円を図形描画クラス1にて描画する。
            {
                //円の外径、内径、頂点数、中心位置を指定。
                var circle = new ace.CircleShape();
                circle.OuterDiameter = 100;
                circle.InnerDiameter = 10;
                circle.NumberOfCorners = 96;
                circle.Position = new ace.Vector2DF(100, 50);

                //円を描画する図形として設定し、合成するテクスチャも設定。
                geometryObj1.Shape = circle;
                geometryObj1.Texture = texture;
                geometryObj1.Position = new ace.Vector2DF(0, 0);
            }

            //弧を図形描画クラス2にて描画する。
            {
                //弧の外径、内径、頂点数、中心位置、開始頂点番号、終了頂点番号を指定。
                var arc = new ace.ArcShape();
                arc.OuterDiameter= 100;
                arc.InnerDiameter = 10;
                arc.NumberOfCorners = 96;
                arc.Position = new ace.Vector2DF(300, 50);
                arc.StartingCorner = 90;
                arc.EndingCorner = 5;

                //弧を描画する図形として設定し、合成するテクスチャも設定。
                geometryObj2.Shape = arc;
                geometryObj2.Texture = texture;
                geometryObj2.Position = new ace.Vector2DF(0, 0);
            }

            //線分を図形描画クラス3にて描画する。
            {
                //線分の始点、終点、太さを指定。
                var line = new ace.LineShape();
                line.StartingPosition = new ace.Vector2DF(410, 50);
                line.EndingPosition = new ace.Vector2DF(630, 50);
                line.Thickness = 5.0f;

                //線分を描画する図形として設定。
                geometryObj3.Shape = line;
                geometryObj3.Position = new ace.Vector2DF(0, 0);
            }

            //矩形を図形描画クラス4にて描画する。
            {
                //矩形の描画範囲、UV範囲を指定。
                var rect = new ace.RectangleShape();
                rect.DrawingArea = new ace.RectF(10, 110, 300, 200);
                rect.UV = new ace.RectF(0, 0, 0.5f, 0.5f);

                //矩形を描画する図形として設定し、合成するテクスチャも設定。
                geometryObj4.Shape = rect;
                geometryObj4.Texture = texture;
                geometryObj4.Position = new ace.Vector2DF(0, 0);
            }

            //三角形を図形描画クラス5にて描画する。
            {
                //三角形の各頂点の位置とUVを指定。
                var triangle = new ace.TriangleShape();
                triangle.SetPointByIndex(new ace.Vector2DF(320, 350), 0);
                triangle.SetPointByIndex(new ace.Vector2DF(100, 450), 1);
                triangle.SetPointByIndex(new ace.Vector2DF(540, 450), 2);

                triangle.SetUVByIndex(new ace.Vector2DF(0.5f, 0.2f), 0);
                triangle.SetUVByIndex(new ace.Vector2DF(0.1f, 0.5f), 1);
                triangle.SetUVByIndex(new ace.Vector2DF(0.9f, 0.7f), 2);

                //三角形を描画する図形として設定し、合成するテクスチャも設定。
                geometryObj5.Shape = triangle;
                geometryObj5.Texture=texture;
                geometryObj5.Position= new ace.Vector2DF(0, 0);
            }

            //多角形を図形描画クラス6にて描画する。
            {

                var polygon = new ace.PolygonShape();
                //多角形を構成する頂点を追加していく。（星形になるようにする。）
                for (int i = 0; i < 10; ++i)
                {
                    ace.Vector2DF vec = new ace.Vector2DF(1, 0);
                    vec.Degree = i * 36;
                    vec.Length = (i % 2 == 0) ? 100 : 55;
                    polygon.AddVertex(vec + new ace.Vector2DF(500, 250));

                }

                //多角形を描画する図形として設定し、合成するテクスチャも設定。
                geometryObj6.Shape = polygon;
                geometryObj6.Texture = texture;
                geometryObj6.Position = new ace.Vector2DF(0, 0);
            }

            // aceが進行可能かチェックする。
            while (ace.Engine.DoEvents())
            {
                // aceを更新する。
                ace.Engine.Update();
            }

            // aceを終了する。
            ace.Engine.Terminate();
        }

    }
}